lib/sshutils/sftp/parse.go

```diff
@@ -17,64 +17,119 @@ limitations under the License.
-		`(?:\[[^@\[\]]+\])` +
-		// or
-		`|` +
-		// some stuff without brackets or colons to make sure the OR condition
-		// is not ambiguous
-		`(?:[^@\[\:\]]+)` +
-		`)` +
-		// after colon, there is a path that could consist technically of
-		// any char including empty which stands for the implicit home directory
-		`:(?P<path>.*)`,
-)
-
-// Destination is SCP destination to copy to or from
+// Destination is a remote SFTP destination to copy to or from.
 type Destination struct {
 	// Login is an optional login username
 	Login string
 	// Host is a host to copy to/from
-	Host utils.NetAddr
+	Host *utils.NetAddr
 	// Path is a path to copy to/from.
 	// An empty path name is valid, and it refers to the user's default directory (usually
 	// the user's home directory).
 	// See https://tools.ietf.org/html/draft-ietf-secsh-filexfer-09#page-14, 'File Names'
 	Path string
 }
 
-// ParseSCPDestination takes a string representing a remote resource for SFTP
-// to download/upload, like "user@host:/path/to/resource.txt" and parses it into
+// ParseDestination takes a string representing a remote resource for SFTP
+// to download/upload in the form "[user@]host:[path]" and parses it into
 // a structured form.
 //
 // See https://tools.ietf.org/html/draft-ietf-secsh-filexfer-09#page-14, 'File Names'
 // section about details on file names.
-func ParseDestination(s string) (*Destination, error) {
-	out := reSFTP.FindStringSubmatch(s)
-	if len(out) < 4 {
-		return nil, trace.BadParameter("failed to parse %q, try form user@host:/path", s)
+func ParseDestination(input string) (*Destination, error) {
+	firstColonIdx := strings.Index(input, ":")
+	if firstColonIdx == -1 {
+		return nil, trace.BadParameter("%q is missing a path, use form [user@]host:[path]", input)
 	}
-	addr, err := utils.ParseAddr(out[2])
-	if err != nil {
-		return nil, trace.Wrap(err)
+	hostStartIdx := strings.LastIndex(input[:firstColonIdx], "@")
+	// if a login exists and the path begins right after the login ends,
+	// no host is specified
+	if hostStartIdx != -1 && hostStartIdx+1 == firstColonIdx {
+		return nil, trace.BadParameter("%q is missing a host, use form [user@]host:[path]", input)
+	}
+
+	var login string
+	// If at least one '@' exists and is before the first ':', get the
+	// login. Otherwise, either there are no '@' or all '@' are after
+	// the first ':' (where the host or path starts), so no login was
+	// specified.
+	if hostStartIdx != -1 {
+		login = input[:hostStartIdx]
+		// increment so that we won't try to parse the host starting at '@'
+		hostStartIdx++
+	} else {
+		hostStartIdx = 0
+	}
+
+	pathStartIdx := firstColonIdx + 1
+	var host *utils.NetAddr
+	// if the host begins with '[', it is most likely an IPv6 address,
+	// so attempt to parse it as such
+	afterLogin := input[hostStartIdx:]
+	if afterLogin[0] == '[' {
+		ipv6Host, hostEndIdx, err := parseIPv6Host(input, hostStartIdx)
```

Comment:
I see the ipv6 logic is being handled here now, but wont that make the path validation above invalid?

Reply:
I'm not sure what you mean, the only path validation I do above this is returning an error if no colons are in the input at all.

Reply:
Consider the following example: `foo@[ff:ff:...]`  There a user and host are provided, but no path / file is provided.  For this reason above the error should be thrown, but it will be considered valid due to matching the `:` in the host.

Reply:
Ah I see good catch, let me fix that

Reply:
Wait I actually already tested that: https://github.com/gravitational/teleport/blob/cd1e012d9df7efe712528c9aea8173e862d05f0b/lib/sshutils/sftp/parse_test.go#L229-L233

You can see where I check for that in `parseIPv6Host`